notebooks/working/investigating_deconfounder.py

```diff
@@ -0,0 +1,592 @@
+
+# +
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+from causalgraphicalmodels import CausalGraphicalModel, StructuralCausalModel
+import pylogit
+from collections import OrderedDict
+import pylogit as cm
+from functools import reduce
+import statsmodels.api as sm
+import statsmodels.formula.api as smf
+from math import ceil
+
+from IPython import display
+import seaborn as sns
+
+import tensorflow as tf
+import numpy as np
+import numpy.random as npr
+import pandas as pd
+import tensorflow as tf
+import tensorflow_probability as tfp
+import statsmodels.api as sm
+
+from tensorflow_probability import edward2 as ed
+from sklearn.datasets import load_breast_cancer
+from pandas.plotting import scatter_matrix
+from scipy import sparse, stats
+from sklearn.model_selection import train_test_split
+from sklearn.linear_model import LogisticRegression 
+from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score, roc_curve
+
+import matplotlib
+matplotlib.rcParams.update({'font.sans-serif' : 'Helvetica',
+                            'axes.labelsize': 10,
+                            'xtick.labelsize' : 6,
+                            'ytick.labelsize' : 6,
+                            'axes.titlesize' : 10})
+import matplotlib.pyplot as plt
+
+import seaborn as sns
+color_names = ["windows blue",
+               "amber",
+               "crimson",
+               "faded green",
+               "dusty purple",
+               "greyish"]
+colors = sns.xkcd_palette(color_names)
+sns.set(style="white", palette=sns.xkcd_palette(color_names), color_codes = False)
+
+
+import os
+os.listdir('.')
+
+# +
+drive_alone_graph = CausalGraphicalModel(
+    nodes=['a', 'b', 'c', 'confounder',
+          'd',  'e', 'f', 'y'],
+    edges=[
+        ("a", "y"), 
+        ("b", "y"), 
+        ("c", "y"), 
+        ("d", "y"), 
+        ("e", "y"), 
+        ("f", "y"), 
+        ("confounder", "y"), 
+        
+        
+        ("confounder", "a"),
+        ("confounder", "b"),
+#         ("confounder", "c"),
+        ("confounder", "d"),
+        ("confounder", "e"),
+        ("confounder", "f"),
+
+
+        
+        
+        ("confounder", "y"), 
+        ("confounder", "y"), 
+    ]
+)
+
+# draw return a graphviz `dot` object, which jupyter can render
+drive_alone_graph.draw()
+
+# +
+sample_size = 10000
+
+confounder = np.random.normal(loc=20, scale = 5, size = sample_size)
+
+# +
+a = .3*confounder + np.random.normal(loc=10,scale=3,size=sample_size)
+b = -1*confounder + np.random.normal(loc=6,scale=3,size=sample_size)
+c = np.random.normal(loc=15,scale=3,size=sample_size) # + .5*confounder 
+d = .8*confounder + np.random.normal(loc=-10,scale=3,size=sample_size)
+e = .5*confounder + np.random.normal(loc=8,scale=3,size=sample_size)
+f= -.3*confounder + np.random.normal(loc=-12,scale=3,size=sample_size)
+y = 5 + 2*a + 2*b - 5*c + 3*d + 2*e - 5*f + 7*confounder + np.random.normal(loc=0,scale=1,size=sample_size)
```

Comment:
So far, in the simulation shown in this notebook, you
- simulate 1 value of the confounder
- simulate 1 corresponding choice observation
- try to recover the confounder from that 1 choice observation

Based on plots of inferred vs actual confounders, my guess would be we are not at all recovering the confounders. Basically, we don't have enough information to identify (i.e. infer with any reasonable accuracy) the confounders.

Could you try re-running the simulation where you simulate multiple choice observations per individual with the same value of the latent confounder across all choice situations?

I imagine that as the number of observations per individual increases, our ability to recover the confounder for each individual will increase, and then the deconfounded estimates in the outcome model will more closely match the data generating parameters.